File: invalidParanthesis.java
// Time Complexity : O(2^n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach

class Solution {
    public List<String> removeInvalidParentheses(String s) {
        List<String> result = new ArrayList<>();
        HashSet<String> set = new HashSet<>();
        Queue<String> q =  new LinkedList<>();
        boolean flag = false;
        q.add(s);
        while(!q.isEmpty() && !flag)
        {
            int size = q.size();
            for(int i = 0 ; i <size;i++)
            {
                String currString = q.poll();
                if(isValid(currString))
                {
                    result.add(currString);
                    flag = true;
                    
                }
                if(!flag)
                {
                    for(int x = 0 ; x < currString.length();x++)
                    {
                        if(Character.isLetter(currString.charAt(x))) continue; 
                        String child = currString.substring(0,x) + currString.substring(x+1);
                        if(!set.contains(child))
                        {
                            set.add(child);
                            q.add(child);
                        }
                
                   } 
                }

                
            }
        }
        return result;
    }
    private boolean isValid(String s)
    {
        int count =0;
        for(Character c:s.toCharArray())
        {
            if(c == '(')
            {
                count++;
            }
            else if (c==')')
            {
                if(count==0) return false;
                count--;
            }
        }
        return count==0;
    }
}

File: InValidParanthesis.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace DFS_And_BFS
{
    public class InValidParanthesis
    {
        /*
         * T.C: O(2 to the power N) 
         * S.C: O(N) since we are saving all the combination in the queue
         */
        public IList<string> RemoveInvalidParentheses(string s)
        {
            if (s == null || s.Length == 0) return null;

            List<string> result = new List<string>();
            int max = 0;
            HashSet<string> set = new HashSet<string>();
            //BFS
            Queue<string> queue = new Queue<string>();
            set.Add(s);
            queue.Enqueue(s);
            bool flag = false;

            while (queue.Count != 0 && flag == false)
            {

                string curr = queue.Dequeue();
                if (isValid(curr))
                {
                    set.Add(curr);
                    result.Add(curr);
                    flag = true;
                }
                else
                {
                    for (int j = 0; j < curr.Length; j++)
                    {
                        if (Char.IsLetter(curr[j])) continue;

                        string str = curr.Substring(0, j) + curr.Substring(j + 1, curr.Length - 1);

                        if (!set.Contains(str))
                        {
                            set.Add(str);
                            queue.Enqueue(str);
                        }
                    }
                }

            }


            return result;

        }

        private bool isValid(string s)
        {
            int count = 0;
            for (int i = 0; i < s.Length; i++)
            {
                if (s[i] == '(')
                {
                    count++;
                }
                else if (s[i] == ')')
                {
                    count--;
                }
            }

            return count == 0;
        }
    }
}
